libs/gl-plugin/src/tramp.rs

```diff
@@ -488,15 +454,99 @@ async fn reestablished_channels(
             Ok(_amount) => Some(channel_data),
             _ => None,
         })
-        .collect::<Vec<ChannelData>>())
+        .collect::<Vec<Channel>>())
 }
 
-struct ChannelData {
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Channel {
     short_channel_id: cln_rpc::primitives::ShortChannelId,
     spendable_msat: u64,
     min_htlc_out_msat: u64,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct ChannelContribution<'a> {
+    channel: &'a Channel,
+    contrib_msat: u64,
+}
+
+/// Finds an allocation that covers the target amount using as few channels as
+/// possible.
+fn find_minimal_allocation<'a>(
+    channels: &'a [Channel],
+    target_msat: u64,
+) -> Option<Vec<ChannelContribution<'a>>> {
+    // We can not allocate channels for a zero amount.
+    if target_msat == 0 {
+        return None;
+    }
+
+    // We'll use recursion with backtracking.
+    // - `idx` is the current channel index.
+    // - `target_msat` is the remaining amount.
+    // - `current` collects the allocations so far.
+    // - `best` stores the best solution found.
+    fn rec<'a>(
+        channels: &'a [Channel],
+        idx: usize,
+        target_msat: u64,
+        current: &mut Vec<ChannelContribution<'a>>,
+        best: &mut Option<Vec<ChannelContribution<'a>>>,
+    ) {
+        // Base case: If we've exactly allocated the target, record the solution.
+        if target_msat == 0 {
+            if best.is_none() || current.len() < best.as_ref().unwrap().len() {
+                *best = Some(current.clone())
+            }
+            return;
+        }
+
+        // Base case: We have processed all channels and could not find a
+        // solution.
+        if idx >= channels.len() {
+            return;
+        }
+
+        // Option 1: Skip the current channel.
+        rec(channels, idx + 1, target_msat, current, best);
```

Comment:
This starts by skipping the largest channels. I think it would be best to start with the ideal case, and the most probable case: the htlc fits in a single channel (the first channel). Basically the base cases that would work most of the time is 'take the maximum from each channel until the amount fits. 

I think the algorithm here should return immediately once 'a' solution has been found. It does not have to be the best solution. 

So skipping the current channel should happen last. (unless channels are ordered from smallest to largest).

Reply:
I was assuming that it might be helpful to fully drain channels to allow for re-allocation or consolidation of those channels, and since we sort the channels: largest spendable msat first, What should happen here is that we prefer to drain smaller channels fist. Choosing the best case instead of just choosing a case that works was part of the same assumtion I made. I can add a test to verify this behavior or we can change to any other policy you prefer.

Reply:
That does sound good actually, to drain the smaller channels first.